### PR TITLE
Integration with Pybullet simulation + fixing json loading functionality!

### PR DESCRIPTION
**Pull request by nkumar-bdai**
This PR makes a number of changes to enable proper bilevel planning with spot environments. The primary change is filling out the `simulate` method of the `SpotEnv`, and all other changes are in service of this!

### CODE REVIEW

File: predicators/envs/spot_env.py [hunk 1]
```diff
@@ -671,7 +752,48 @@ def _get_next_nonpercept_atoms(self, obs: _SpotObservation,
...
+        # believes it has finished the task. Used for goal checking.
+        if action_name == "done":
+            # During planning, trust that the goal is accomplished if the
+            # done action is returned.
+            self._current_task_goal_reached = True
+            return state
+
+        # Execute the action in the pybullet env. Automatically retry
+        # if a retryable error is encountered.
+        action_fn(*action_fn_args)  # type: ignore
```
Comment (tsilver-bdai) [inline predicators/envs/spot_env.py:777]: wait, now I'm very confused. what's the difference between the action function and the simulation function instead the action extra info?

Comment (tsilver-bdai) [reply predicators/envs/spot_env.py:777]: ohhh I see, this is the simulation function. change the name!

File: predicators/envs/spot_env.py [hunk 2]
```diff
@@ -671,7 +752,48 @@ def _get_next_nonpercept_atoms(self, obs: _SpotObservation,
...
+
+        # In the future, we'll probably actually implement proper ways to
+        # check things like InHandView and other quantities below
+        # in the simulator (e.g., by shooting a ray!). However for now,
+        # we're going to just hack things such that the expected atoms check
+        # passes. This will be done in a manner similar to dry sim.
+        if action_name == "MoveToHandViewObject":
+            obj_to_view = action_objs[1]
+            next_state.set(obj_to_view, "in_hand_view", 1.0)
+        elif action_name == "PickObjectFromTop":
```
Comment (tsilver-bdai) [inline predicators/envs/spot_env.py:791]: are these two just the ones that have been implemented so far? should we add an assertion that no other action is being used yet?

File: predicators/envs/spot_env.py [hunk 3]
```diff
@@ -765,10 +887,28 @@ def render_state_plt(
...
+            json_dict, object_name_to_object)
+        # Get the object detection id's, which are features of
+        # each object type.
+        obj_to_detection_id = {
+            v: k
+            for k, v in self._detection_id_to_obj.items()
+        }
+        for obj, init_val in init_dict.items():
+            if obj in obj_to_detection_id:
+                init_val["object_id"] = obj_to_detection_id[
```
Comment (tsilver-bdai) [inline predicators/envs/spot_env.py:904]: what exactly is the type issue here?

File: predicators/ground_truth_models/spot_env/options.py [hunk 4]
```diff
@@ -90,6 +113,67 @@ def _grasp_at_pixel_and_maybe_stow_or_dump(
         stow_arm(robot)
 
 
+def _sim_safe_grasp_at_pixel_and_maybe_stow_or_dump(
+        robot: Robot, target_obj: Object, rng: np.random.Generator,
+        timeout: float, retry_grasp_after_fail: bool, do_stow: bool,
+        do_dump: bool) -> None:
+    """Implicitly gets the current image from the hand camera and selects a
```
Comment (tsilver-bdai) [inline predicators/ground_truth_models/spot_env/options.py:120]: I know this is an unavoidable hack for now, but I do think we can reduce the amount of copied code between here and the function above

**Review by tsilver-bdai — COMMENTED** [review]
really nice work! complicated stuff

File: predicators/envs/spot_env.py [hunk 5]
```diff
@@ -198,6 +223,30 @@ def __init__(self, use_gui: bool = True) -> None:
...
+                # First, launch pybullet.
+                pbrspot.utils.connect(use_gui=True)
+                pbrspot.utils.disable_real_time()
+                pbrspot.utils.set_default_camera()
+                # Create robot object atop the floor.
+                self.sim_robot = pbrspot.spot.Spot()
+                floor_urdf = utils.get_env_asset_path(
+                    "urdf/spot_related/floor.urdf")
+                floor_obj = pbrspot.body.createBody(floor_urdf)
+                # The floor is known to be about 60cm below the
```
[the remark this answers is not in the document]
Comment (nkumar-bdai) [reply predicators/envs/spot_env.py:242]: yup! need to figure that out in the future lol

File: predicators/envs/spot_env.py (hunk 3 above)
Comment (nkumar-bdai) [reply predicators/envs/spot_env.py:907]: the issue is that the `init_dict` expects to be a mapping from objects to a dict of features and floats, but for spot envs, one of the features of an object is an `ObjectDetectionID`, which is not a float

File: predicators/envs/spot_env.py [hunk 6]
```diff
@@ -165,6 +176,20 @@ def get_detection_id_for_object(obj: Object) -> ObjectDetectionID:
     return obj_to_detection_id[obj]
 
 
+def get_simulated_robot() -> Optional[pbrspot.spot.Spot]:
+    """Return the simulated robot object."""
+    return _SIMULATED_SPOT_ROBOT
+
+
+def get_simulated_object(
```
[the remark this answers is not in the document]
Comment (tsilver-bdai) [reply predicators/envs/spot_env.py:182]: @NishanthJKumar I understand now why we want these methods to be global, but why do we want lookup to be allowed to fail?

File: predicators/structs.py [hunk 7]
```diff
@@ -1808,6 +1808,18 @@ def __len__(self) -> int:
...
 
 
+@dataclass(frozen=True, repr=False, eq=False)
+class SpotActionExtraInfo:
+    """A sequence of things that are in the extra_info field of actions for the
+    SpotEnv."""
+    action_name: str
+    operator_objects: Sequence[Object]
+    real_world_fn: Optional[Callable]
+    real_world_fn_args: Sequence
```
Comment (tsilver-bdai) [inline predicators/structs.py:1818]: should we make this a Tuple instead? same for below. for example, you're needing to assert that it's an instance of a tuple in at least one place. if you change the type, you should be able to remove that assertion

**Review by tsilver-bdai — APPROVED** [review]
great work!

**Pull request merged**